src/client/websocket/WebSocketManager.js

```diff
@@ -1,34 +1,79 @@
-const EventEmitter = require('events');
-const { Events, Status } = require('../../util/Constants');
-const WebSocketConnection = require('./WebSocketConnection');
+const WebSocketShard = require('./WebSocketShard');
+const { Events, Status, WSEvents } = require('../../util/Constants');
+const PacketHandlers = require('./handlers');
+
+const BeforeReadyWhitelist = [
+  WSEvents.READY,
+  WSEvents.RESUMED,
+  WSEvents.GUILD_CREATE,
+  WSEvents.GUILD_DELETE,
+  WSEvents.GUILD_MEMBERS_CHUNK,
+  WSEvents.GUILD_MEMBER_ADD,
+  WSEvents.GUILD_MEMBER_REMOVE,
+];
 
 /**
  * WebSocket Manager of the client.
- * @private
  */
-class WebSocketManager extends EventEmitter {
+class WebSocketManager {
   constructor(client) {
-    super();
     /**
      * The client that instantiated this WebSocketManager
      * @type {Client}
      */
     this.client = client;
 
     /**
-     * The WebSocket connection of this manager
-     * @type {?WebSocketConnection}
+     * The gateway this WebSocketManager uses.
+     * @type {string}
+     */
+    this.gateway = undefined;
+
+    /**
+     * An array of shards spawned by this WebSocketManager.
+     * @type {WebSocketShard[]}
+     */
+    this.shards = [];
+
+    /**
+     * An array of queued shards to be spawned by this WebSocketManager.
+     * @type {Array<WebSocketShard|number|string>}
+     */
+    this.spawnQueue = [];
+
+    /**
+     * Whether or not this WebSocketManager is currently spawning shards.
+     * @type {boolean}
+     */
+    this.spawning = false;
+
+    /**
+     * An array of queued events before this WebSocketManager became ready.
+     * @type {Immediate[]}
```

Comment:
I don't think this is an array of `Immediate`s:
See [here](https://github.com/discordjs/discord.js/pull/2902/files#diff-25533af2af0a39d16191a08cc17046b9R169)